monai/transforms.py

```diff
@@ -0,0 +1,62 @@
+
+from monai.transforms import (SpatialPadd, Compose, CropForegroundd, LoadImaged, RandFlipd, 
+            RandCropByPosNegLabeld, Spacingd, RandRotate90d, ToTensord, NormalizeIntensityd, 
+            EnsureType, RandWeightedCropd, HistogramNormalized, EnsureTyped, Invertd, SaveImaged,
+            EnsureChannelFirstd, CenterSpatialCropd, RandSpatialCropSamplesd, Orientationd, 
+            Rand3DElasticd, RandBiasFieldd)
+
+# median image size in voxels - taken from nnUNet
+# median_size = (123, 255, 214)
+# so pad with this size
+
+def train_transforms(crop_size, num_samples_pv, lbl_key="label"):
+    return Compose([   
+            LoadImaged(keys=["image", lbl_key]),
+            EnsureChannelFirstd(keys=["image", lbl_key]),
+            # Orientationd(keys=["image", lbl_key], axcodes="RPI"),
+            Spacingd(keys=["image", "label"], pixdim=(1.0, 1.0, 1.0), mode=("bilinear", "bilinear"),),
+            CropForegroundd(keys=["image", lbl_key], source_key="image"),     # crops >0 values with a bounding box
+            SpatialPadd(keys=["image", lbl_key], spatial_size=(64, 128, 128), method="symmetric"),
+            # SpatialPadd(keys=["image", lbl_key], spatial_size=(123, 255, 214), method="symmetric"),
+            # RandSpatialCropSamplesd(keys=["image", lbl_key], roi_size=crop_size, num_samples=num_samples_pv, random_center=True, random_size=False),
+            RandCropByPosNegLabeld(keys=["image", "label"], label_key="label",
+                                   spatial_size=crop_size, pos=1, neg=1, num_samples=num_samples_pv, 
+                                   # if num_samples=4, then 4 samples/image are randomly generated
+                                   image_key="image", image_threshold=0.),
```

Comment:
This should ensure you always get non-empty batches, which does not seem to be the case (from: https://wandb.ai/whats_ai/contrast-agnostic/runs/bwlz3e92?workspace=user-whats_ai)

![image](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/assets/2482071/484662bd-1166-457e-9db3-3dcbcdd712bc)

Reply:
So apparently always getting non-empty patches (note, not _batches_) is not for training as it biases the model towards input that only have some label. This might not necessarily be the case in validation where we do a sliding window inference and there are empty patches that could be encountered there. 

Hence, what's usually done is "foreground oversampling", where we oversample patches with sure-shot labels (let's call them positive patches). nnUNet does 2/3rds foreground oversampling of the positive patches and the rest are negatives. In a similar vein, I am taking it up a notch and I do 3/4ths foreground oversampling (in [these](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/blob/407ae043c35b88eccb3f73678b9cb0323b23df3f/monai/transforms.py#L22-L27) lines). So, unlike earlier where we did 50-50 sampling of pos/neg patches, the prob. of having positive patches is 75% now. The resulting images can be seen in this [demo run on wandb](https://wandb.ai/naga-karthik/contrast-agnostic/runs/1qpld7dg?workspace=user-naga-karthik).